tokio-buf/src/buf_stream/size_hint.rs

```diff
@@ -0,0 +1,79 @@
+/// A `BufStream` size hint
+///
+/// The default implementation returns:
+///
+/// * 0 for `available`
+/// * 0 for `lower`
+/// * `None` for `upper`.
+#[derive(Debug, Default, Clone)]
+pub struct SizeHint {
+    available: usize,
+    lower: usize,
+    upper: Option<usize>,
```

Comment:
For use in hyper to set the content-length, this being a `usize` *does* mean that 32bit platforms can't set the content-length automatically if sending over 4GBs. Also, things like implementing `BufStream` for `tokio::fs::File` would cause problems there as well, since [`Metadata::len()`](https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.len) returns a `u64`.

Reply:
Changing `size_hint` to `u64` has the interesting property of making `collect()` fallible.